crates/core/src/live_socket/navigation/ffi.rs

```diff
@@ -0,0 +1,285 @@
+/// An action taken with respect to the history stack
+/// when [NavCtx::navigate] is executed. defaults to
+/// Push behavior.
+#[derive(uniffi::Enum, Default, Clone)]
+pub enum NavAction {
+    /// Push the navigation event onto the history stack.
+    #[default]
+    Push,
+    /// Replace the current top of the history stack with this navigation event.
+    Replace,
+}
+
+/// Options for calls to [NavCtx::navigate]
+#[derive(Default, uniffi::Record)]
+pub struct NavOptions {
+    #[uniffi(default = None)]
+    pub action: Option<NavAction>,
+    #[uniffi(default = None)]
+    pub extra_event_info: Option<Vec<u8>>,
+    #[uniffi(default = None)]
+    pub state: Option<Vec<u8>>,
+}
+
+impl NavEvent {
+    pub fn new(
+        event: NavEventType,
+        to: NavHistoryEntry,
+        from: Option<NavHistoryEntry>,
+        info: Option<Vec<u8>>,
+    ) -> Self {
+        let new_url = Url::parse(&to.url).ok();
+        let old_url = from.as_ref().and_then(|dest| Url::parse(&dest.url).ok());
+
+        let same_document = old_url
+            .zip(new_url)
+            .is_some_and(|(old, new)| old.path() == new.path());
+
+        NavEvent {
+            event,
+            same_document,
+            from,
+            to,
+            info,
+        }
+    }
+
+    pub fn new_from_reload(dest: NavHistoryEntry, info: Option<Vec<u8>>) -> NavEvent {
+        NavEvent::new(NavEventType::Reload, dest.clone(), dest.into(), info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::traverse_to] event
+    pub fn new_from_traverse(
+        new_dest: NavHistoryEntry,
+        old_dest: NavHistoryEntry,
+        info: Option<Vec<u8>>,
+    ) -> NavEvent {
+        NavEvent::new(NavEventType::Traverse, new_dest, old_dest.into(), info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::navigate] event
+    pub fn new_from_navigate(
+        new_dest: NavHistoryEntry,
+        old_dest: Option<NavHistoryEntry>,
+        opts: NavOptions,
+    ) -> NavEvent {
+        let event = match opts.action {
+            Some(NavAction::Replace) => NavEventType::Replace,
+            _ => NavEventType::Push,
+        };
+
+        NavEvent::new(event, new_dest, old_dest, opts.extra_event_info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::back] event,
+    /// passing info into the event handler closure.
+    pub fn new_from_forward(
+        new_dest: NavHistoryEntry,
+        old_dest: Option<NavHistoryEntry>,
+        info: Option<Vec<u8>>,
+    ) -> NavEvent {
+        NavEvent::new(NavEventType::Push, new_dest, old_dest, info)
+    }
+
+    /// Create a new nav event from the details of a [NavCtx::back] event,
+    /// passing info into the event handler closure.
+    pub fn new_from_back(
+        new_dest: NavHistoryEntry,
+        old_dest: NavHistoryEntry,
+        info: Option<Vec<u8>>,
+    ) -> NavEvent {
+        NavEvent::new(NavEventType::Push, new_dest, Some(old_dest), info)
+    }
```

Comment:
I wish there was a cleaner way to write these. Like, I want `impl From<T> for NaveEvent` but there's no great way to express that.

Reply:
granted that these are all basically one liners this does seem a little gratuitous on second glance.